user_pill: Use variables for sizing and scale with font-size.

evykassirer opened the pull request:
for squished pills, at 12px, 14px, 16px, 20px:

| before | after |
| --- | --- |
| ![Screen Shot 2025-03-11 at 15 53 04](https://github.com/user-attachments/assets/1525b8a3-a5b7-40b4-add8-a570d20bd459) | ![Screen Shot 2025-03-11 at 16 05 45](https://github.com/user-attachments/assets/5eed9bb9-8eb5-4ecd-90b3-3d145aacf92a) |
| ![Screen Shot 2025-03-11 at 15 53 53](https://github.com/user-attachments/assets/4fe28b2d-1051-4179-8192-a37c1fa0c496) | ![Screen Shot 2025-03-11 at 16 07 09](https://github.com/user-attachments/assets/f323bace-a014-4476-98df-bed7d05af409) |
| ![Screen Shot 2025-03-11 at 15 54 26](https://github.com/user-attachments/assets/17cff3e0-a433-4545-8be7-38d2347d948c) | ![Screen Shot 2025-03-11 at 16 04 23](https://github.com/user-attachments/assets/54b9c544-0324-46d3-bb88-fdba68b42239) |
| ![Screen Shot 2025-03-11 at 15 54 55](https://github.com/user-attachments/assets/2d998ed9-a1c6-4d56-a853-3cf491c6e80f) | ![Screen Shot 2025-03-11 at 16 00 29](https://github.com/user-attachments/assets/d0ea5997-a022-4e8b-b01e-a951ced89cff) |


for pills with full width, at 12px, 14px, 16px, 20px:

| before | after |
| --- | --- |
| ![Screen Shot 2025-03-11 at 16 10 51](https://github.com/user-attachments/assets/26ca59cb-6de5-4083-972c-4043c41e0609) | ![Screen Shot 2025-03-11 at 16 09 03](https://github.com/user-attachments/assets/a26433ec-1224-484e-a1b5-26a7673b6ce9) |
| ![Screen Shot 2025-03-11 at 16 10 40](https://github.com/user-attachments/assets/29cfa0b4-de31-4d59-84a6-1ea10544e801) | ![Screen Shot 2025-03-11 at 16 08 46](https://github.com/user-attachments/assets/14448b08-d260-4932-90ba-3d774d53cb62) |
| ![Screen Shot 2025-03-11 at 16 10 30](https://github.com/user-attachments/assets/bafeb569-9ba4-4f85-b038-c2bd0e3ad85a) | ![Screen Shot 2025-03-11 at 16 09 19](https://github.com/user-attachments/assets/b4256138-5132-4c26-bfd4-4d17169ac50e) |
| ![Screen Shot 2025-03-11 at 16 10 19](https://github.com/user-attachments/assets/cd52006c-824b-4ed6-a079-87f224de829a) | ![Screen Shot 2025-03-11 at 16 09 38](https://github.com/user-attachments/assets/1d7bb02d-ced4-4ad5-973d-dc5dbed94af0) |

alya commented:
Hm, I don't love all that extra space in the last screenshot. Can we prevent the pill from abbreviating in the first screenshot without making it overly long in the last?

karlstolley commented on web/styles/search.css:
```diff
@@ -290,11 +290,15 @@
 
     .pill-label {
-        min-width: 30px;
+        min-width: 2.5em; /* 35px at 14px/em */
```
Have you experimented with something like [fit-content](https://developer.mozilla.org/en-US/docs/Web/CSS/fit-content)? Possibly coupled with `min()`? The Zoe example that @alya pointed out reveals the limits of trying to eyeball a min-width here.

evykassirer replied:
Ah, I'd forgotten `fit-content` could be used with `min`, I was trying to do something like that yesterday. Thanks!

evykassirer replied:
<img width="496" alt="image" src="https://github.com/user-attachments/assets/22f6a962-21e2-480c-89ef-fe30e4ae9416" />

Looks like that's not something that can be done?

evykassirer replied:
I'm not sure if there's a way to do the thing we want, but maybe you have further ideas? I can also come back to this once dealing with some other PR feedback.

karlstolley replied:
@evykassirer, I just put up #33951, which solves a similar problem to this. My approach there was to rework how flex items were allowed to flex, and brought a crafty `max-width` to the table, too. Maybe have a look at that, and see if the same basic idea can't be brought to pills here?

evykassirer replied:
That feels different enough. The flex stuff you're doing there is already present on the pill labels. The problem you're solving has a clear max-width, but here the problem is that we don't want to have a `min-width`. But without a `min-width`, how does it know where to cut off pills?

evykassirer replied:
I want to note that we had a `min-width` and occasional right padding previously, and here I'm just making it a little wider. Here's an example on CZO at 16px:

<img width="198" alt="image" src="https://github.com/user-attachments/assets/2f4d9bc1-2353-4f10-973e-f238fbeffb5c" />

@alya since there's not an easy solution, how would you feel about merging this and opening a followup issue to explore potential better solutions? I can also make the min-width for the pill label a bit more narrow than it's set in this PR. Do you have a preference for the width?

_(with screenshots at 14px, and I would change it so it scales with font-size)_

30px -- what it was already before this PR

<img width="622" alt="image" src="https://github.com/user-attachments/assets/be1efb04-5965-4350-b444-c86599b31f16" />

<img width="352" alt="image" src="https://github.com/user-attachments/assets/8f38d057-8927-4a88-b903-68a71112bf09" />

35px -- what I changed it to, essentially allowing 3 characters instead of 2, with the caveat that short names have more right padding

<img width="604" alt="image" src="https://github.com/user-attachments/assets/8b466f6f-dc57-442a-bce6-0aa714a7be2c" />

<img width="386" alt="image" src="https://github.com/user-attachments/assets/05eb2c20-3a08-4f23-a354-a7e0f52a4792" />

evykassirer replied:
I'm also noticing descenders are getting cut off, so the pills need to be a bit taller. I can put up a PR for that.

evykassirer commented:
Updated the PR and updated screenshots added above. Still feeling torn about the tradeoff of more letters in the cut off pill vs the extra space after short names. I still think more letters is the better option, since most names are much longer than "Zoe".

alya commented:
I still don't like Zoe having all that extra space. Might still be worth it to fix the narrow screens situation, but I'd really prefer a solution that preserves this as-is.


![image](https://github.com/user-attachments/assets/231d3b14-81e2-4b19-9c13-a94f53d39261)

evykassirer commented:
That screenshot is only showing this happening in the 20px font-size case. You can see in the "before" screenshots I posted yesterday that there's already space after Zoe at 14px. So that's what would be "as is". There was already a min-width set.

One difference is the min-width wasn't scaling with font-size, and in this PR I changed it to scale with font-size. I think it's better to be consistent than to have that trailing space at some font-sizes and not others. We should decide what that minimum width is though.

I can also post in frontend and see if anyone has ideas.

alya commented:
Sure, posting in #frontend seems useful to see if we can avoid it.

evykassirer commented:
https://chat.zulip.org/#narrow/channel/6-frontend/topic/something.20like.20min.28MIN_WIDTH.2C.20fit-content.29/with/2121288

evykassirer commented:
I think this is promising!

It would be nice if Zoe was wider in a case like this:

<img width="571" alt="image" src="https://github.com/user-attachments/assets/f28afa35-4abc-4e4c-a05f-5174516a1c5b" />

At some widths the `...` isn't appearing, but that was also an issue with the fix I currently have up.

<img width="417" alt="image" src="https://github.com/user-attachments/assets/c825b4a5-9c34-4d7f-98c7-032f806b6405" />

I think either of these solutions is better than the current state of things even if they're not perfect, and probably with your changes is better? I'll add them to my branch and push it so @alya can play around with it.

karlstolley commented:
Yeah. This is probably calling for some refactoring post-10.0 of the numerous flexboxes in the area. Presumably the `dm:` pill should be wider to accommodate that piece of UI, but ideally we'd want usernames getting to share the same amount of space (Cordelia shouldn't get more space because a) she's in the `dm:` pill or b) because she has a longer username.

alya commented:
I left a comment [on CZO](https://chat.zulip.org/#narrow/channel/9-issues/topic/search.20pills.20can.20get.20squashed/near/2126197)[#issues > search pills can get squashed @ 💬](https://chat.zulip.org/#narrow/channel/9-issues/topic/search.20pills.20can.20get.20squashed/near/2126197) about one more tweak if possible, but I think it's fine to merge this and go from here.
